syntax = "proto3";

option java_multiple_files = true;
option java_package = "br.ufu.facom.gbc074.projeto.database";

package database;

message Usuario{
  // CPF do usuario (chave)
  string cpf     = 1;
  string nome    = 2;
  // campo presente apenas no portal biblioteca
  bool bloqueado = 3;
}

message Livro {
  // ISBN do livro (chave)
  string isbn    = 1;
  string titulo  = 2;
  string autor   = 3;
  int32 total    = 4;
  // campo presente apenas no portal biblioteca
  int32 restante = 5;
}

message Status {
  // 0 = sucesso, 1 = erro
  int32 status = 1; 
  // detalhes do erro para status = 1
  string msg   = 2;
}

message Identificador {
  // cpf para usuario, isbn para livro
  string id = 1;
}

message UsuarioLivro {
  Identificador usuario = 1;
  Identificador livro   = 2;
}

message UsuarioBloqueado {
  Usuario usuario       = 1;
  // livro(s) que causou(aram) bloqueio
  repeated Livro livros = 2;
}

message Criterio {
  string criterio = 1;
}

message Vazia {}

message TableMessage {
	string tName = 1;
}

message CreateUpdateMessage {
	string tName = 1;
	string key = 2;
	string value = 3;
}

message RequestMessage {
	string tName = 1;
	string key = 2;
}

message RequestReturn {
	string value = 1;
}

service DatabaseService {
	rpc CreateTable(TableMessage) returns (Status) {}
  rpc DeleteTable(TableMessage) returns (Status) {}
	rpc	Create(CreateUpdateMessage) returns (Status) {}
	rpc	Request(RequestMessage) returns (RequestReturn) {}
  rpc RequestAll(TableMessage) returns (stream RequestReturn) {}
  rpc	Update(CreateUpdateMessage) returns (Status) {}
  rpc	Delete(RequestMessage) returns (Status) {}
}